### ISSUE TITLE
Help in use with SSLH, Advanced SSH, and ProxyJump, using the SISH Docker Image

### ISSUE THREAD

**Issue by shadowrylander** [antoniomika/sish#82]
Hello! <br>
&nbsp;&nbsp;&nbsp;&nbsp; I wish to use `sish` with `sslh` and ProxyJumps, but I can't seem to figure out how to set this up; what I'd like to do is ssh into my `Linode` VPS, caught by `sslh` on ports 443 and 80, and using `sish`, jump to my Raspbery Pi 3 at home, then jump to my primary server, a laptop.
<br> &nbsp;&nbsp;&nbsp;&nbsp; Any help with setting such a system up will be greatly appreciated!

**Comment by antoniomika** [comment]
Right, so things would work like this:
![Untitled drawing](https://user-images.githubusercontent.com/6239308/78604797-577c2480-7828-11ea-9cf7-213499e21d48.png)
Blue arrow can be thought of how you would connect to your raspberry pi from your laptop. Being unfamiliar with inlets, I hope that is what you would expect.

**Comment by shadowrylander** [comment]
Yep! Then from the RPi3, I want to jump to another system on my internal network. So how does the connection get to the RPi3 from the Linode relay?

**Comment by antoniomika** [comment]
If that's the case, you'd probably want to do a double proxyjump (you'd have to use your ssh config as part of this). I'd imagine it would look like this:
`~/.ssh/config`:
```
Host rpi
    ProxyJump sish.linode
```
Then you could do `ssh -J rpi someotherhostonyourlan.local`. Is that what you are referring to?

**Comment by shadowrylander** [comment]
Ah; now I understand. What exactly is the `sish` server?  And what would happen if it were to go down for any reason? Sorry, still confused how the Linode and RPi3 connect without running anything on the latter.

I use [`ASSH`](https://github.com/moul/assh) to manage my connections, so the jumping should be easy enough.

**Comment by antoniomika** [comment]
`sish` is the program you run from this repository. It is an SSH session multiplexer that handles TCP and HTTP forwarding transiently. Think nginx but for ssh tunnels.

**Comment by shadowrylander** [comment]
Fair; but nginx still requires you to go from, say in my case, the RPi3 to the Linode, and not / can't go the other way around; and to reverse tunnel, wouldn't my Pi need to [call] the Linode first, before the Linode connects my laptop to the Pi? If nothing is running on the Pi that calls to the Linode, how is it being connected?

**Comment by antoniomika** [comment]
Correct. You need to run an SSH command on the PI to establish the tunnel. If you take a look at the diagram I sent, that should show you an example command that would work.

**Comment by shadowrylander** [comment]
Oh! Sorry, the fact that the black line between the RPi and the Linode was an arrow got lost with the blue ones. So would you recommend, say, using autossh or mosh to keep the connection alive, and setup a script for when the device reboots?

**Comment by antoniomika** [comment]
Exactly, I would recommend using autossh to establish the tunnel. There are many tutorials online for using systemd or upstart to keep and manage the process. sish sadly won't work with mosh

**Comment by shadowrylander** [comment]
Would that be due to the ProxyJumps? If so, I think the `--experimental-remote-ip=remote` option of mosh, wrapped in [`assh wrapper ssh`](https://github.com/moul/assh#register-the-wrapper-optional) would work! Courtesy of this: [`Investigate what can be done with/by/using/for mosh`](https://github.com/moul/assh/issues/129#issuecomment-429193788)

Regarding keeping the ssh reverse tunnel service alive on the Pi, I'm thinking of going with `runit` instead!

**Comment by antoniomika** [comment]
`runit` should work too! And the reason sish won't work with mosh is because it only implements the SSH protocol, not the mosh protocol, which could be a problem :)

**Comment by shadowrylander** [comment]
Sigh... No way to implement it? :joy_cat: So finally, my setup is this:
* SSH to Linode on Ports 443 / 80
* `SSLH` forwards the packets to `SISH` on Port 2222, to keep 22 open for administrative purposes
* `SISH` uses the reverse tunnel from the RPi3 to tunnel back to my home network
* Jump from the RPi3 to my primary server, a Manjaro Laptop

Problem solved! Thanks so much for the help!

**Comment by antoniomika** [comment]
Awesome glad it worked out! Re: mosh, I wanted to implement it but didn't have time to implement the mosh protocol when I was writing this. The go stdlib has great default protocol implementations (ssh included) which made this fairly trivial to implement. I'll definitely have this in the back of my mind though!

**Comment by shadowrylander** [comment]
Great! Is there an open issue or request I can subscribe to for more news on that?

**Comment by shadowrylander** [comment]
Okay, so it turns out I still can't connect to, at the very least, to my RPi3; the command I'm using is this:
`ssh -p 2222 -J jumper@rteii.duckdns.org:443 rpi3`
I've got the Pi's SSHD listening on Ports 22 and 2222, the reverse tunnels from the Pi to the Linode seem to be up (or at least, no errors have been reported), and I've got SSLH on the Linode listening on Ports 443 and 80, where sslh-select uses the former, and sslh-fork use the latter to connect to the internal sshd port, i.e. 22. I used docker compose to set up sish with port 2222 in the container bound to Linode's port 2222, so when sslh-select gets an ssh packet, it should theoretically pass it on to Linode's port 2222, which should pass it on to the container's port 2222, which should pass it on to the reverse tunnel from the Pi.

Also, sorry for the rambling tone; it's 6 A.M. here. :joy_cat:

**Comment by shadowrylander** [comment]
@shadowrylander Regarding the second port I'm binding to in the reverse tunnel command, i.e. for `localhost`, should that be my Linode's address, i.e. `rteii.duckdns.org`?

**Comment by shadowrylander** [comment]
Here's my docker-compose, by the way (`DOMAIN` is set to `rteii.duckdns.org`):
```yml
# From: https://github.com/antoniomika/sish/blob/master/deploy/docker-compose.yml

version: '3.7'

services:
  busybox:
    image: busybox
    volumes:
      - ./letsencrypt:/etc/letsencrypt
    command: /bin/sh -c "echo '$DOMAIN *.$DOMAIN autorestart-containers=sish' > /etc/ letsencrypt/domains.conf"
  letsencrypt:
    image: adferrand/letsencrypt-dns:latest
    container_name: letsencrypt-dns
    depends_on:
      - busybox
    environment:
      - VERSION=latest
      - LETSENCRYPT_USER_MAIL=titaniumfiles@outlook.com
      - LEXICON_PROVIDER=cloudflare
      - LEXICON_PROVIDER_OPTIONS="--auth-username=titaniumfiles@outlook.com --auth-token=[REDACTED]"
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
      - ./letsencrypt:/etc/letsencrypt
    restart: always
  sish:
    image: antoniomika/sish:latest
    container_name: sish
    depends_on:
      - letsencrypt
    volumes:
      - ./letsencrypt:/etc/letsencrypt
      - ./pubkeys:/pubkeys
      - ./keys:/keys
    command: |
      -sish.addr=2222:2222
      -sish.https=8443:8443
      -sish.http=9080:9080
      -sish.httpsenabled=true
      -sish.httpspems=/etc/letsencrypt/live/$DOMAIN
      -sish.keysdir=/pubkeys
      -sish.pkloc=/keys/ssh_key
      -sish.bindrandom=false
      -sish.domain=$DOMAIN
      -sish.forcerandomsubdomain=false
    network_mode: host
    restart: always
```

VERY IMPORTANT QUESTION:
What template do I use for the Cloudflare API, and what options do I set for it?

**Comment by shadowrylander** [comment]
In addition, I realized I never mentioned that my Linode's address is set to `rteii.duckdns.org`; here is the `runit` script I'm using:
```bash
#!/bin/sh
export USER="jumper"
exec 2>&1

# From: [SSH TUNNELLING FOR FUN AND PROFIT: AUTOSSH](https://www.everythingcli.org/ssh-tunnelling-for-fun-and-profit-autossh/)
runuser -l $USER -c '/usr/bin/fghack autossh -fM 0 -o "ServerAliveInterval 30" -o "ServerAliveCountMax 3" -R rpi3:2222:localhost:2222 jumper@rteii.duckdns.org'
runuser -l $USER -c '/usr/bin/fghack autossh -fM 0 -o "ServerAliveInterval 30" -o "ServerAliveCountMax 3" -R rpi3:9080:localhost:9080 jumper@rteii.duckdns.org'
runuser -l $USER -c '/usr/bin/fghack autossh -fM 0 -o "ServerAliveInterval 30" -o "ServerAliveCountMax 3" -R rpi3:8443:localhost:8443 jumper@rteii.duckdns.org'
```
Basically, since they weren't working in the foreground normally, which `runit` requires, I pushed them to the background using autossh's `-f` option, then brought them back to the foreground using daemontools' `/usr/bin/fghack`. Not sure if that's an issue or not.

**Comment by antoniomika** [comment]
Shouldn't your RPI connect to port 443/2222 on your linode box? The runit script doesn't seem to do that? As for the cloudflare api, you don't have to use that unless you want to have ssl support. The template that I use in my docker-compose are all you need and you can grab the api key using [this documentation](https://support.cloudflare.com/hc/en-us/articles/200167836-Managing-API-Tokens-and-Keys). You should also use `localhost` for the SSH command (as you already have it). Another thing to keep in mind, you can chain multiple `-R` forwards to SSH, so you can use one single runuser command for all 3 port forwards

**Comment by shadowrylander** [comment]
1. The modified `runit` script on `rpi3`, where `2222` is the ssh port in the `sish` container:
```bash
#!/bin/sh
export USER="jumper"
exec 2>&1

# From: [SSH TUNNELLING FOR FUN AND PROFIT: AUTOSSH](https://www.everythingcli.org/ssh-tunnelling-for-fun-and-profit-autossh/)
runuser -l $USER -c '/usr/bin/fghack autossh -fM 0 -o "ServerAliveInterval 30" -o "ServerAliveCountMax 3" -p 2222 \
        -R rpi3:2222:localhost:2222 jumper@rteii.duckdns.org \
        -R rpi3:9080:localhost:9080 jumper@rteii.duckdns.org \
        -R rpi3:8443:localhost:8443 jumper@rteii.duckdns.org \'
```
2. Will any options / permissions work for the API, or are there any specifics I need?
3. Regarding `localhost`: yeah, I though as much. Thanks!
4. This is the output of `docker-compose -f [path] ps --all`:
```
     Name                    Command                                 State        Ports
--------------------------------------------------------------------------------------------
letsencrypt-dns          /run.sh                                      Up
sish                            /app/sish -sish.addr=2222: ...   Restarting
sish_busybox_1        /bin/sh -c echo 'rteii.duc ...        Exit 1
```

**Comment by shadowrylander** [comment]
Since `443` is the port `sslh-select` uses to catch and demultiplex incoming requests, and then send them on to the appropriate port bound to the `sish` container, should _all_ my ssh connections go through there, including the reverse tunnels from the `rpi3`?

**Comment by antoniomika** [comment]
Correct, all of you connections should be to that port. sish handles both types of connections (reverse tunnels and forward tunnels). You also don't need the `jumper@rteii.duckdns.org` on each line with your forward request. It can just be something like 
```
-R ... \
-R ... \
-R ... \
jumper@rteii.duckdns.org
```

**Comment by shadowrylander** [comment]
Right; I forgot that the remote address is only to be used once. Blame the vim yank function. :joy_cat: 
1. Modified `runit` script:
```bash
#!/bin/sh
export USER="jumper"
exec 2>&1

# From: [SSH TUNNELLING FOR FUN AND PROFIT: AUTOSSH](https://www.everythingcli.org/ssh-tunnelling-for-fun-and-profit-autossh/)
runuser -l $USER -c '/usr/bin/fghack autossh -fM 0 -o "ServerAliveInterval 30" -o "ServerAliveCountMax 3" -p 443 \
        -R rpi3:2222:localhost:2222 \
        -R rpi3:9080:localhost:9080 \
        -R rpi3:8443:localhost:8443 \
jumper@rteii.duckdns.org \'
```
2. I'm assuming the command to use to jump to `rpi3` through the Linode would then be the following:
```bash
ssh -p 2222 -J jumper@rteii.duckdns.org:443 rpi3
```

**Comment by shadowrylander** [comment]
Still not working; for the jump I used ports `22`, `2222`, and `443`, but none work.

**Comment by shadowrylander** [comment]
Should I show you my systemd `sshd` logs for the `rpi3` and Linode?

**Comment by antoniomika** [comment]
To jump to the pi, you would need to use an ssh config for the proxyjump machine (since it's not running on port 22). You'd have to set it to use `Port 443` as part of the config entry and then you would use that for the proxy jump

**Comment by shadowrylander** [comment]
Thanks for the log notice!
Why does my using the jump command `-J` not work? I though it was the `ssh` equivalent of the config's `ProxyJump`? In addition, the `proxyjump` machine you're referring to is the Linode, right?

**Comment by antoniomika** [comment]
Yes, Linode is the machine I'm referring to. The reason the ProxyJump doesn't work as is is due to the use of a non-standard port. That effectively prevents the proxyjump from working.

**Comment by shadowrylander** [comment]
Ah; that makes sense. Even without the ports specified in the command itself?
If that's the case, could you possibly write me up a small example, like what the config on the Linode and the `rpi3` would look like? I'm still quite new to this. How would the config reference the `sish` container, if at all?

**Comment by antoniomika** [comment]
So you would not need any configuration for Linode or the rpi. As long as your Linode is running the sish container and your rpi is connected to sish (running on port 443, being multiplexed by sslh) you're all set there. On your laptop trying to get to the rpi, you would want something like this:
```
Host rpi3
    ProxyJump linode

Host linode
    Port 443
```
Then, you would login to the rpi just by doing `ssh rpi3`.

The sish container is what you are connecting to when you're logging into your Linode from the laptop and the rpi reverse tunnel. It's using the SSH protocol, but it has nothing to do with sshd.

**Comment by shadowrylander** [comment]
If you are familiar with the `assh` yaml syntax, this is what I'm using:
```yaml
hosts:
    # Jump Hosts
    jarvis:
            User: jumper
            Port: 443
            Inherits:
                    - argus

    # Shadow Hosts
    rpi3:
            Hostname: 192.168.0.64
            Aliases:
                    - r
            Gateways:
                    # - direct
                    - jarvis
    argus:
            Aliases:
                    - a
            Hostname: rteii.duckdns.org
            Port: 80

defaults:
        Port: 22
        User: shadowrylander
```
Where `argus` is the Linode, `jarvis` is the version of the `argus` settings I use to jump to the `rpi3` (for security reasons), and `rpi3` is my final destination; the defaults are Ports 22, and my user.

**Comment by shadowrylander** [comment]
Unfortunately, it still doesn't work; it just hangs, and doesn't even give me an error. The errors seem to be in the logs, however.

**Comment by shadowrylander** [comment]
My `ufw` firewall on the `rpi3` includes `2222`, `9080`, and `8443`, while the Linode includes `443`, and `80`.
Also, oh... I just realized I should probably change the port `rpi3` uses in the `assh` config... Just a moment, please!

**Comment by shadowrylander** [comment]
> My `ufw` firewall on the `rpi3` includes `2222`, `9080`, and `8443`, while the Linode includes `443`, and `80`.
> Also, oh... I just realized I should probably change the port `rpi3` uses in the `assh` config... Just a moment, please!

Changed it, and still doesn't seem to work. Sigh...

**Comment by antoniomika** [comment]
Are you referring to the sish logs or the sshd logs? You shouldn't be seeing anything in sshd any this could be a sign that your sslh setup isn't working correctly.

**Comment by shadowrylander** [comment]
I was referring to the sshd logs; but the reason I think `sslh` is working fine is because I can get in to my Linode using port `80` just fine, which `sslh-fork` redirects to the Linode's port `22`, which is where it's own sshd is listening. Is there any reason `sslh-select`, listening on port `443`, could be failing to pass on the packet to `sish`, such as incorrect configuration of the container? My `docker-compose.yml` is above, and these are my `sslh` configs:

`sslh-fork.cfg:`
```cfg
# Default Arch configuration
# You can find more examples in /usr/share/doc/sslh

verbose: false;
foreground: true;
inetd: false;
numeric: false;
transparent: false;
timeout: 2;
user: "root";

listen:
(
    { host: "0.0.0.0"; port: "80"; }
);

protocols:
(
     { name: "ssh"; service: "ssh"; host: "localhost"; port: "22"; probe: "builtin"; },
     { name: "openvpn"; host: "localhost"; port: "1194"; probe: "builtin"; },
     { name: "xmpp"; host: "localhost"; port: "5222"; probe: "builtin"; },
     { name: "http"; host: "localhost"; port: "9080"; probe: "builtin"; },
     { name: "ssl"; host: "localhost"; port: "8443"; probe: "builtin"; },
     { name: "anyprot"; host: "localhost"; port: "8443"; probe: "builtin"; }
);

# vim:set ts=4 sw=4 et:
```

`sslh-select.cfg:`
```cfg
# Default Arch configuration
# You can find more examples in /usr/share/doc/sslh

verbose: false;
foreground: true;
inetd: false;
numeric: false;
transparent: false;
timeout: 2;
user: "root";

listen:
(
    { host: "0.0.0.0"; port: "443"; }
);

protocols:
(
     { name: "ssh"; service: "ssh"; host: "localhost"; port: "2222"; probe: "builtin"; },
     { name: "openvpn"; host: "localhost"; port: "1194"; probe: "builtin"; },
     { name: "xmpp"; host: "localhost"; port: "5222"; probe: "builtin"; },
     { name: "http"; host: "localhost"; port: "9080"; probe: "builtin"; },
     { name: "ssl"; host: "localhost"; port: "8443"; probe: "builtin"; },
     { name: "anyprot"; host: "localhost"; port: "8443"; probe: "builtin"; }
);

# vim:set ts=4 sw=4 et:
```

Or do you want to see my `systemd` configs and logs for the two `sslh` setups as well?

**Comment by antoniomika** [comment]
Are you running sslh in docker? Now that I think about it, sslh may not be able to connect to sish because docker does some iptables magic in order to handle port proxying. If that's the case, you'll want to get the sish container ip (from docker inspect) and use that as the host parameter

**Comment by shadowrylander** [comment]
Ah; no, I'm not running sslh in a docker container as well. So to reiterate, I should run docker inspect on the sish container and use that in the sslh configs? Wonder if I could use a variable in the latter...

**Comment by shadowrylander** [comment]
Oh, and side note: I've noticed that, while the `letsencrypt` container has been up for quite some time now, the `sish` keeps saying it's restarting for some reason. I think that's why I can't seem to get an IP from it.

**Comment by shadowrylander** [comment]
```bash
Last login: Wed Apr  8 07:45:48 2020 from ::1
Linux localhost 5.5.13-arch2-1 x86_64 unknown
λ docker ps                                                                   (~) 07:49:31CONTAINER ID        IMAGE                              COMMAND                  CREATED             STATUS                          PORTS               NAMES
4c1193c680e2        antoniomika/sish:latest            "/app/sish -sish.add…"   25 hours ago        Restarting (1) 56 seconds ago                       sish
88ce9d77295f        adferrand/letsencrypt-dns:latest   "/run.sh"                25 hours ago        Up 7 hours                                          letsencrypt-dns
λ docker ps                                                                   (~) 07:49:35CONTAINER ID        IMAGE                              COMMAND                  CREATED             STATUS                          PORTS               NAMES
4c1193c680e2        antoniomika/sish:latest            "/app/sish -sish.add…"   25 hours ago        Restarting (1) 24 seconds ago                       sish
88ce9d77295f        adferrand/letsencrypt-dns:latest   "/run.sh"                25 hours ago        Up 7 hours                                          letsencrypt-dns
λ                                                                             (~) 07:50:04
```

**Comment by antoniomika** [comment]
That would also be the problem! What are the logs from sish? You can check via `docker logs -f sish`

**Comment by shadowrylander** [comment]
```
2020/04/07 07:07:30 Starting SSH service on address: 2222:2222
2020/04/07 07:07:30 listen tcp: lookup 2222: no such host
2020/04/07 07:07:31 Starting SSH service on address: 2222:2222
2020/04/07 07:07:31 listen tcp: lookup 2222: no such host
```
And that's it, repeated to infinity (and beyond :joy_cat: ).

**Comment by shadowrylander** [comment]
I thought that `2222:2222` was referring to a `host:container` port configuration?

**Comment by shadowrylander** [comment]
Wait: changed it to use `:2222`, and this is cropping up repeatedly:
```
2020/04/08 19:04:05 Starting SSH service on address: :2222
2020/04/08 19:04:05 open /etc/letsencrypt/live/rteii.duckdns.org/fullchain.pem: no such file or directory
2020/04/08 19:04:06 Starting SSH service on address: :2222
2020/04/08 19:04:06 open /etc/letsencrypt/live/rteii.duckdns.org/fullchain.pem: no such file or directory
```

**Comment by shadowrylander** [comment]
Do I perhaps need a new `Auth Token` from Cloudflare?

**Comment by antoniomika** [comment]
Potentially. What are the logs you see from cloudflare?

**Comment by shadowrylander** [comment]
How would I do that again...?

**Comment by antoniomika** [comment]
Just the logs from letsencrypt: `docker logs -f letsencrypt-dns`

**Comment by shadowrylander** [comment]
Dammit; already deleted the token. I'll create a new one and try again, but I still need to know the minimum permissions to use.

**Comment by antoniomika** [comment]
Sorry to say, but I don't have that information. Here's the only documentation that I know of for that: https://dnsrobocert.readthedocs.io/en/latest/providers_options.html#cloudflare

**Comment by shadowrylander** [comment]
Sigh... Fair enough. I'll get back to you with the status of the container once I update the Auth Token!

**Comment by shadowrylander** [comment]
Here are the results of `docker logs -f letsencrypt-dns`:
```
2020-04-08 21:29:50 81f1fd853d8d dnsrobocert.core.main[1] INFO Starting DNSroboCert.
2020-04-08 21:29:50 81f1fd853d8d dnsrobocert.core.main[1] INFO Configuration file is in draft mode: no action will be done.
```

**Comment by shadowrylander** [comment]
@antoniomika 'Ello...?

**Comment by antoniomika** [comment]
Hi @shadowrylander, sorry for not getting back to you. The past few weeks have been a bit crazy for me. It looks like the version of letsencrypt I used when I first included that in the repo has drastically changed. I'll go through and update that. I'm tracking that issue in this general issue: #85. I'm gonna close this as it's not really an issue with sish, but I'd be happy to keep helping over the closed issue or over IRC.
